### PR TITLE
Add hardhat as an explicit option in udapp

### PR DESCRIPTION
**Pull request by Aniket-Engg**

### CODE REVIEW

File: apps/remix-ide/src/app/tabs/runTab/settings.js [hunk 1]
```diff
@@ -244,16 +244,29 @@ class SettingsUI {
...
   setExecutionContext (context) {
-    this.blockchain.changeExecutionContext(context, () => {
-      modalDialogCustom.prompt('External node request', this.web3ProviderDialogBody(), 'http://127.0.0.1:8545', (target) => {
-        this.blockchain.setProviderFromEndpoint(target, context, (alertMsg) => {
-          if (alertMsg) addTooltip(alertMsg)
-          this.setFinalContext()
-        })
+    if (context === 'Hardhat Provider') {
+      this.blockchain.changeExecutionContext(context, () => {
+        modalDialogCustom.prompt('Hardhat node request', this.hardhatProviderDialogBody(), 'http://127.0.0.1:8545', (target) => {
```
Comment (yann300) [inline apps/remix-ide/src/app/tabs/runTab/settings.js:249]: wasn't it possible to put this in https://github.com/ethereum/remix-project/pull/1206/files#diff-aedcee9b7a11a19d2ec1b9a1831bd5d4985392e3701b74003c0171f1b2e74c2bR21 ?
I mean your way work but it'd be better to have hardhat stuff in a single.
Also having declared a whole plugin but still need to do that is a bit overkill.

File: apps/remix-ide/src/app/tabs/hardhat-provider.js [hunk 2]
```diff
@@ -58,7 +55,8 @@ const sendAsyncInternal = (provider, data, resolve, reject) => {
       resolve(message)
     })
   } else {
-    resolve({ jsonrpc: '2.0', result: [], id: data.id })
+    const result = data.method === 'net_listening' ? 'canceled' : []
```
Comment (yann300) [inline apps/remix-ide/src/app/tabs/hardhat-provider.js:58]: actually maybe a better way would have been to `reject` if the provider is not set..

File: apps/remix-ide/src/blockchain/execution-context.js [hunk 3]
```diff
@@ -189,14 +188,16 @@ export class ExecutionContext {
...
-      if (!err && isConnected) {
+      if (!err && isConnected === true) {
         this.executionContext = context
         this._updateBlockGasLimit()
         this.event.trigger('contextChanged', [context])
         this.event.trigger('web3EndpointChanged')
         cb()
+      } else if (isConnected === 'canceled') {
+        web3.setProvider(oldProvider)
+        cb()
```
Comment (yann300) [inline apps/remix-ide/src/blockchain/execution-context.js:200]: yes, indeed that's a clever solution.
Perhaps `cb()` should also return an `error` (same as in the `else` below)

**Review by yann300 — APPROVED** [review]

File: apps/remix-ide/src/app/tabs/hardhat-provider.js (hunk 2 above)
Comment (Aniket-Engg) [reply apps/remix-ide/src/app/tabs/hardhat-provider.js:63]: `reject` will also show a predefined error on toaster

File: apps/remix-ide/src/blockchain/execution-context.js (hunk 3 above)
Comment (Aniket-Engg) [reply apps/remix-ide/src/blockchain/execution-context.js:200]: Generally, no message is shown on clicking on `cancel` in a modal

File: apps/remix-ide/src/app/tabs/runTab/settings.js (hunk 1 above)
Comment (Aniket-Engg) [reply apps/remix-ide/src/app/tabs/runTab/settings.js]: done

**Review by yann300 — APPROVED** [review]

**Pull request merged**